Enable 64-bit support for Mapbox

yanokwa opened the pull request:
Closes #2844 

This PR includes the 64-bit Mapbox native code. My expectations are that with this change:
1. Mapbox runs on 32-bit and 64-bit Android devices
1. Mapbox continues to gracefully fail on all x86 devices
1. The Collect APK file size increases because of this native code

Something that caught my eye is that the x86 lib ends in`.so1` instead of `.so`. I have confirmed that the production APK doesn't have any x86 libs, so I think the extension is correct.

yanokwa committed:
Enable 64-bit support for Mapbox

shobhitagarwal1612 commented:
@yanokwa I'm curious why `.so1` was used as the native binary's extension as I have only seen and compiled `.so` files fo far. Moreover, the source code also [loads](https://github.com/mapbox/mapbox-gl-native/blob/f66848efaf1efe2ce4088bf56ae7af3a765b9d57/platform/android/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/LibraryLoader.java#L43) the file with `.so` extension.

seadowg commented:
Disclaimer: There is a lot of *thinking* in this post that could certainly be wrong.

@shobhitagarwal1612 I *think* it could be to do with versioning in dynamic languages - it lets you have two versions of the lib sitting next to each other (a `.so` and a `.so1`). I also *think* that `System.load` won't care about the extension as it's just looking for the lib name.

seadowg commented on collect_app/build.gradle:
```diff
@@ -151,10 +151,9 @@ android {
-        exclude 'lib/x86/libmapbox-gl.so'
+        // the Mapbox native library only for 32-bit and 64-bit ARM devices and
+        // omit it for all X86 devices.
+        exclude 'lib/x86/libmapbox-gl.so1'
```
So when I build the APK (after commenting out all our excludes for mapbox libs) and analyze it this is what I see for the x86 libs:

<img width="1105" alt="Screenshot 2019-07-05 at 17 14 12" src="https://user-images.githubusercontent.com/556280/60734598-699efe80-9f48-11e9-936b-be60a3a39318.png">

Were you seeing a `libmapbox-gl.so1`? I tried both the `debug` and `odkCollectRelease` build types just in case it was related to that.

yanokwa replied:
I don't remember seeing `libmapbox-g1.so1` in any of the APKs. What I did verify was unzipping the latest beta and confirming that we had the armeabi-v7a lib. 
<img width="539" alt="Screen Shot 2019-07-05 at 9 24 43 AM" src="https://user-images.githubusercontent.com/32369/60735055-c24da780-9f06-11e9-85ff-d28c16986c4c.png">

seadowg replied:
Right but the code was previously excluding `lib/x86/libmapbox-gl.so`. If I build an APK from this PR I'm seeing this in libs:

<img width="1124" alt="Screenshot 2019-07-05 at 17 46 06" src="https://user-images.githubusercontent.com/556280/60735901-f8ae1580-9f4c-11e9-8825-b450f0778445.png">

I thought you were seeing an `so1` file in there so you changed the extension to that in the `exclude`. Sorry if I'm totally misunderstanding what's going on here 😕.

lognaturel commented:
@yanokwa, @seadowg's [check](https://github.com/opendatakit/collect/pull/3191#discussion_r300747541) looks to me like it conclusively shows the lib should not end in `so1`. I'm wondering whether you just ended up with a typo. You write "I have confirmed that the production APK doesn't have any x86 libs, so I think the extension is correct" but the change to `so1` is introduced by this PR so as @shobhitagarwal1612 notes, there should actually be an explanation for why it got changed.

yanokwa commented:
I introduced the change `.so1` change, but didn't notice. Then I wasted everyone's time by misread the discussion. Apologies, all!

I have reverted the change and force-pushed. I have confirmed that the PR produces an APK with the ARM libs.

<img width="825" alt="Screen Shot 2019-07-08 at 3 39 37 PM" src="https://user-images.githubusercontent.com/32369/60847164-a2341780-a196-11e9-9d7e-4b0c29c9645b.png">

mmarciniak90 commented:
**Tested with success**

**Verified on Android:** 8.1
Verified also other android versions to check possible regression: 7.0, 6.0, 5.1, 4.4
I was not able to use the device with Android 4.2 because of the crash reported in #3192 

**Verified cases:**
* all Geo widgets with Mapbox (used different Basemaps)
* Geo widgets with Google Maps and OSM
* changing maps during filling form
* set map before starting the form
* import/export setting with Mapbox
* 3 input methods

@opendatakit-bot unlabel "needs testing"
@opendatakit-bot label "behavior verified"
